matrix_content_scanner/servlets/__init__.py

```diff
@@ -0,0 +1,249 @@
+#  Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import abc
+import json
+import logging
+from typing import Any, Awaitable, Callable, Optional, Tuple
+
+from twisted.internet import defer
+from twisted.web.http import Request
+from twisted.web.resource import Resource
+from twisted.web.server import NOT_DONE_YET
+
+from matrix_content_scanner import logutils
+from matrix_content_scanner.crypto import CryptoHandler
+from matrix_content_scanner.utils.constants import ErrCode
+from matrix_content_scanner.utils.encrypted_file_metadata import (
+    validate_encrypted_file_metadata,
+)
+from matrix_content_scanner.utils.errors import ContentScannerRestError
+from matrix_content_scanner.utils.types import JsonDict
+
+logger = logging.getLogger(__name__)
+
+
+class _AsyncResource(Resource, metaclass=abc.ABCMeta):
+    def render(self, request: Request) -> int:
+        """This gets called by twisted every time someone sends us a request."""
+        defer.ensureDeferred(self._async_render(request))
+        return NOT_DONE_YET
+
+    async def _async_render(self, request: Request) -> None:
+        """Processes the incoming request asynchronously and handles errors."""
+        # Treat HEAD requests as GET requests.
+        request_method = request.method.decode("ascii")
+        if request_method == "HEAD":
+            request_method = "GET"
+
+        # Set the request type in the logging context.
+        assert request.path is not None
+        parts = request.path.split(b"/")
+        # Paths in the content scanner API use the form
+        # "/_matrix/media_proxy/unstable/{requestType}/...", so the request type is at
+        # index 4 in the parts.
+        logutils.set_request_type_in_context(parts[4].decode("ascii"))
```

Comment:
shouldn't we check that the first 3 parts are what we expect?

Reply:
We _could_, but I'm not sure how useful this would be, given the resource tree building (https://github.com/matrix-org/matrix-content-scanner-python-wip/blob/main/matrix_content_scanner/httpserver.py) ensures that all incoming requests that make it this far use the path `/_matrix/media_proxy/unstable`.

Reply:
I'd probably veer towards an assertion, then. This is the kind of assumption that bites back hard if it ever fails